### ISSUE TITLE
AF 3.2 carshes with Arithmetic exception on Linux

### ISSUE THREAD

**Issue by unbornchikken** [arrayfire/arrayfire#1122]
Donno if it's a duplicate or not, I haven't found similar issue there yet. Today I've upgraded ArrayFire.js to support 3.2. Windows is fine, but on Linux, my tests crashes as soon as I create a new array with an Arithmetic exception. Created a screenshot, when you can see the issue and the stack trace.

http://imagizer.imageshack.us/a/img903/7073/AErESV.png

**Comment by shehzan10** [comment]
I don't see any information about the exception code. I can't see the full line of arrayfire code either. Can you provide the relevant line of code as well as the full exception being generated.

**Comment by unbornchikken** [comment]
Hi, I've debuged firther.

The cause of the crash is that:

https://github.com/arrayfire/arrayfire/blob/devel/src/backend/cpu/platform.cpp#L118

On my computer, default case selected, and mNumSMT stays on zero, so I get divide by zero at:

https://github.com/arrayfire/arrayfire/blob/devel/src/backend/cpu/platform.cpp#L121

**Comment by shehzan10** [comment]
Yeah this was an experimental CPU information tool we added in 3.2. It seems like this part of fetching CPU information is more fragmented than we expected (doesn't work on arm/ppc etc).

@pavanky maybe we should add a runtime variable that disables this in 3.2.1.

**Comment by unbornchikken** [comment]
How should I disable that check? Are there any workarounds?

**Comment by shehzan10** [comment]
Are you using the installer or building arrayfire from source? If you are building from source, you could comment out the relevant parts in [getInfo()](https://github.com/arrayfire/arrayfire/blob/devel/src/backend/cpu/platform.cpp#L194).

If you are using the installer, there isn't much you can do at this point. getInfo will always get called. It's like the initialization function.

**Comment by 9prady9** [comment]
@unbornchikken Can you tell me the CPU this code is running on ?

**Comment by unbornchikken** [comment]
I can live with 3.1 latest till this issue gets fixed.

I'm on my old ThinkPad R500 with:

```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 23
model name      : Intel(R) Core(TM)2 Duo CPU     T6570  @ 2.10GHz
stepping        : 10
microcode       : 0xa0c
cpu MHz         : 2101.000
cache size      : 2048 KB
physical id     : 0
siblings        : 2
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx lm constant_tsc arch_perfmon pebs bts nopl aperfmperf pni dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm sse4_1 xsave lahf_lm ida dtherm tpr_shadow vnmi flexpriority
bogomips        : 4189.33
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:
```

**Comment by 9prady9** [comment]
@unbornchikken thank you. Will try to get my hands one of those or a similar one.

As a temporary fix, we can default the value of `mNumSMT` to one instead of zero in the constructor initialization. That should get rid of the divide by zero exception.

**Comment by unbornchikken** [comment]
That would do definitely. Which is the last version of AF release that is not affected by this? Unfortunatelly I get that crash with 3.1.3 too.

**Comment by 9prady9** [comment]
I believe it is 3.1.0 in which CPUID based code first landed into a release. So, to avoid this issue on this CPU, you need to revert back to 3.0.2.

**Comment by unbornchikken** [comment]
Since I wanted to add and test ArrayFire 3.2 support to ArrayFire.js, I'm gonna build a 3.2 from source with that mNumSMT(1) fix applyed. Thanks for your help.

**Comment by 9prady9** [comment]
Sure, since 3.2.1 is around the corner(30th Nov), you can stick with 3.2.  We will try to push a fix for this in 3.2.1.

**Comment by pavanky** [comment]
Fixed in hotfixes-3.2.1

**Comment by unbornchikken** [comment]
Thanks. Keep up the good work, this library is still one of the most awesome on Github!
